consensus/istanbul/backend/handler.go

```diff
@@ -286,3 +301,175 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
+			return isValidator, nil
+		}
+	}
+}
+
+// shouldSendValidatorProof determines if this node should send a
+// validator proof revealing its address to a peer
+func (sb *Backend) shouldSendValidatorProof(peer consensus.Peer) (bool, error) {
+	var validatorAddress common.Address
+	if sb.config.Proxy {
+		validatorAddress = sb.config.ProxiedValidatorAddress
+	} else {
+		validatorAddress = sb.Address()
+	}
+	// Check to see if this node is a validator
+	regAndActiveVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		return false, err
+	}
+	if !regAndActiveVals[validatorAddress] {
+		return false, nil
+	}
+	// If the peer is not a known validator, we do not want to give up extra
+	// information about ourselves
+	if _, err := sb.valEnodeTable.GetAddressFromNodeID(peer.Node().ID()); err != nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// readValidatorProofMessage reads a validator proof message as a part of the
+// istanbul handshake. Returns if the peer is a validator or if an error occurred.
+func (sb *Backend) readValidatorProofMessage(peer consensus.Peer) (bool, error) {
+	logger := sb.logger.New("func", "readValidatorProofMessage")
+	peerMsg, err := peer.ReadMsg()
+	if err != nil {
+		return false, err
+	}
+
+	var payload []byte
+	if err := peerMsg.Decode(&payload); err != nil {
+		return false, err
+	}
+
+	var msg istanbul.Message
+	err = msg.FromPayload(payload, sb.verifyValidatorProofMessage)
+	if err != nil {
+		return false, err
+	}
+	// If the Signature is empty, the peer has decided not to reveal its info
+	if len(msg.Signature) == 0 {
+		return false, nil
+	}
+
+	var versionedEnode versionedEnode
+	err = rlp.DecodeBytes(msg.Msg, &versionedEnode)
+	if err != nil {
+		return false, err
+	}
+
+	node, err := enode.ParseV4(versionedEnode.EnodeURL)
+	if err != nil {
+		return false, err
+	}
+
+	// Ensure the node in the versionedEnode matches the peer node
+	if node.ID() != peer.Node().ID() {
+		logger.Warn("Peer provided incorrect node ID in versionedEnode", "versionedEnode enode url", versionedEnode.EnodeURL, "peer enode url", peer.Node().URLv4())
+		return false, errors.New("Incorrect node in versionedEnode")
+	}
+
+	// Check if the peer is within the registered/elected valset
+	regAndActiveVals, err := sb.retrieveMaybeStaleRegisteredAndElectedValidators()
+	if err != nil {
+		logger.Trace("Error in retrieving registered/elected valset", "err", err)
+		return false, err
+	}
+	if !regAndActiveVals[msg.Address] {
+		logger.Debug("Received a validator proof message from peer with address that is not a registered or active validator", "msg.Address", msg.Address)
+		return false, nil
+	}
+
+	// If the versionedEnode message is too old, we don't count this proof as valid
+	// An error is given if the entry doesn't exist, so we ignore the error.
+	knownVersion, err := sb.valEnodeTable.GetVersionFromAddress(msg.Address)
+	if err == nil && versionedEnode.Version < knownVersion {
+		return false, nil
```

Comment:
I would like to see a `Debug` log here because this should not happen under normal conditions and would be worth noting. It may even be part of an attempted security violation, much like the use of an expired TLS certificate, but I cannot immediately think of a reasonable attack vector as long as this check is in place.